ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -712,29 +754,40 @@ IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, long operationTimeMs)
    * @return the {@link IndexValue} of the ttl update record
    * @throws StoreException if there is any problem writing the index record
    */
-  private IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, MessageInfo info, long operationTimeMs)
-      throws StoreException {
+  private IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, MessageInfo info, long operationTimeMs,
+      short lifeVersion) throws StoreException {
     validateFileSpan(fileSpan, true);
+    boolean hasLifeVersion = lifeVersion != -1;
+    //TODO IndexValue findLatest
     IndexValue value = findKey(id);
     if (value == null && info == null) {
       throw new StoreException("Id " + id + " not present in index " + dataDir, StoreErrorCodes.ID_Not_Found);
-    } else if (value != null && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+    }
+    short retrievedLifeVersion = value == null ? info.getLifeVersion() : value.getLifeVersion();
+    if (value != null && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
       throw new StoreException("Id " + id + " deleted in index " + dataDir, StoreErrorCodes.ID_Deleted);
     } else if (value != null && value.isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
       throw new StoreException("TTL of " + id + " already updated in index" + dataDir, StoreErrorCodes.Already_Updated);
+    } else if (hasLifeVersion && retrievedLifeVersion > lifeVersion) {
+      throw new StoreException("LifeVersion conflict in index. Id " + id + " LifeVersion: " + retrievedLifeVersion
+          + " Undelete LifeVersion: " + lifeVersion, StoreErrorCodes.Life_Version_Conflict);
     }
     long size = fileSpan.getEndOffset().getOffset() - fileSpan.getStartOffset().getOffset();
     IndexValue newValue;
+
     if (value == null) {
+      lifeVersion = hasLifeVersion ? lifeVersion : info.getLifeVersion();
       // It is possible that the PUT has been cleaned by compaction
+      // but the TTL update is going to still be placed?
       newValue =
-          new IndexValue(size, fileSpan.getStartOffset(), info.getExpirationTimeInMs(), info.getOperationTimeMs(),
-              info.getAccountId(), info.getContainerId());
+          new IndexValue(size, fileSpan.getStartOffset(), IndexValue.FLAGS_DEFAULT_VALUE, info.getExpirationTimeInMs(),
+              info.getOperationTimeMs(), info.getAccountId(), info.getContainerId(), lifeVersion);
       newValue.clearOriginalMessageOffset();
     } else {
+      lifeVersion = hasLifeVersion ? lifeVersion : value.getLifeVersion();
```

Comment:
Undelete version should be increased based  on preview D version. Is it correct to use value.getLifeVersion()? 
Also, since P0   D0   U5  is possible(from replication), we can't rely on value.getLifeVersion().   Use info.getLifeVersion() would be safe.

Reply:
info could be null.  Replication calls to this method should be giving a not-negative-one lifeVersion value, so that will be used (negative-one lifeVersions would only be used for client calls)

Reply:
Got you.